activestorage/app/controllers/active_storage/disk_controller.rb

```diff
@@ -17,6 +17,20 @@ def show
     head :not_found
   end
 
+  def show_public
+    key = params[:key]
+
+    if blob = ActiveStorage::Blob.find_by(key: key)
+      if blob.service.public_service?
+        serve_file disk_service.path_for(blob.key), content_type: blob.content_type, disposition: :inline
+      else
+        head :unauthorized
+      end
+    else
+      head :not_found
+    end
+  end
```

Comment:
First, I’d like to see this moved to a separate controller: `ActiveStorage::Public::DiskController#show`.

`serve_file` is general enough that it could move to Action Controller eventually, but for now we can just move it to `ActiveStorage::BaseController` so `ActiveStorage::DiskController` and `ActiveStorage::Public::DiskController` can share it.

Second, you can also inline the `key` local variable:

```ruby
def show
  if blob = ActiveStorage::Blob.find_by(key: params[:key])
    if blob.service.public_service?
      serve_file disk_service.path_for(blob.key), content_type: blob.content_type, disposition: :inline
    else
      head :unauthorized
    end
  else
    head :not_found
  end
end
```

Third, this action should honor the blob’s potentially-custom disk service by using `blob.service` in place of `disk_service`.

Fourth, I’d like to add `ActiveStorage::Blob#public?` and use it here:

```ruby
class ActiveStorage::Blob
  def public?
    service.public?
  end
end
```

Reply:
I've refactored by splitting the public controller portion to `ActiveStorage::PublicDiskController`. I didn't see an advantage in placing this controller in another level of namespace nesting.